stdlib/universe/exponential_moving_average.go

```diff
@@ -0,0 +1,576 @@
+			case flux.TUInt:
+				err = t.doUInt(cr.UInts(j), builder, j, doExponentialMovingAverage[j])
+			case flux.TFloat:
+				err = t.doFloat(cr.Floats(j), builder, j, doExponentialMovingAverage[j])
+			case flux.TString:
+				err = t.passThroughString(cr.Strings(j), builder, j)
+			case flux.TTime:
+				err = t.passThroughTime(cr.Times(j), builder, j)
+			}
+
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func (t *exponentialMovingAverageTransformation) UpdateWatermark(id execute.DatasetID, mark execute.Time) error {
+	return t.d.UpdateWatermark(mark)
+}
+
+func (t *exponentialMovingAverageTransformation) UpdateProcessingTime(id execute.DatasetID, pt execute.Time) error {
+	return t.d.UpdateProcessingTime(pt)
+}
+
+func (t *exponentialMovingAverageTransformation) Finish(id execute.DatasetID, err error) {
+	t.d.Finish(err)
+}
+
+func (t *exponentialMovingAverageTransformation) passThroughBool(vs *array.Boolean, b execute.TableBuilder, bj int) error {
+	if int64(vs.Len()) < t.n {
+		if vs.IsNull(vs.Len() - 1) {
+			if err := b.AppendNil(bj); err != nil {
+				return err
+			}
+		} else {
+			if err := b.AppendBool(bj, vs.Value(vs.Len()-1)); err != nil {
+				return err
+			}
+		}
+		return nil
+	} else {
+		s := arrow.BoolSlice(vs, int(t.n-1), vs.Len())
+		defer s.Release()
+		return b.AppendBools(bj, s)
+	}
+}
+
+func (t *exponentialMovingAverageTransformation) doInt(vs *array.Int64, b execute.TableBuilder, bj int, doExponentialMovingAverage bool) error {
+	mult := 2.0 / (float64(t.n) + 1)
+	j := 0
+
+	if t.i[bj] == 0 {
+		if vs.IsValid(j) {
+			t.pValue[bj] = float64(vs.Value(j))
+			t.count[bj]++
+		}
+
+		t.i[bj]++
+		j++
+	}
+
+	for ; int64(t.i[bj]) < t.n && j < vs.Len(); t.i[bj]++ {
+		if !vs.IsNull(j) {
+			t.pValue[bj] += float64(vs.Value(j))
+			t.count[bj]++
+		}
+		j++
+	}
+
+	if int64(t.i[bj]) == t.n {
+		if t.count[bj] != 0 {
+			t.pValue[bj] = t.pValue[bj] / t.count[bj]
+			if err := b.AppendFloat(bj, t.pValue[bj]); err != nil {
+				return err
+			}
+		} else {
+			if err := b.AppendNil(bj); err != nil {
+				return err
+			}
+		}
+	}
+
+	l := vs.Len()
+	for ; j < l; j++ {
+		if !doExponentialMovingAverage {
+			if vs.IsValid(j) {
+				if err := b.AppendInt(bj, vs.Value(j)); err != nil {
+					return err
+				}
+			} else {
+				if err := b.AppendNil(bj); err != nil {
+					return err
+				}
+			}
+		} else if vs.IsNull(j) {
+			if t.count[bj] == 0 {
+				if err := b.AppendNil(bj); err != nil {
+					return err
+				}
+			} else {
+				if err := b.AppendFloat(bj, t.pValue[bj]); err != nil {
+					return err
+				}
+			}
+		} else {
+			cValue := float64(vs.Value(j))
+			var ema float64
+			if t.count[bj] == 0 {
+				ema = cValue
+				t.count[bj]++
+			} else {
+				ema = (cValue * mult) + (t.pValue[bj] * (1.0 - mult))
+			}
+			if err := b.AppendFloat(bj, ema); err != nil {
+				return err
+			}
+			t.pValue[bj] = ema
+		}
+		t.i[bj]++
+	}
+	if int64(t.i[bj]) < t.n-1 {
```

Comment:
Is this wrong? If the period is longer than the number of items in the chunk, this evaluates to true